tcp_proxy: fix allowing multiple readDisable calls

ohadvano opened the pull request:
Additional Description: Fixes #27516. For TCP tunneling scenario it's possible that ``readDisable()`` will be called multiple times, either from codec layer and connection layer. This change only asserts in case TCP tunneling is not enabled, and also increases counters only if read was actually disabled/enabled.
Risk Level: low
Testing: Unit tests
Docs Changes: changelog
Release Notes: none
Platform Specific Features: none

ohadvano commented:
/assign @KBaichoo could you please have a first look before I'm proceeding to implementing tests, etc.. Thanks

ohadvano commented:
/assign @KBaichoo
@KBaichoo Could you please have a first look before I'm proceeding to implementing tests, etc? Thanks

ohadvano commented:
/retest

KBaichoo commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -343,7 +350,11 @@ void Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark() {
-  ASSERT(on_high_watermark_called_);
+  // TCP Tunneling may call on high watermark multiple times.
+  if (!parent_->config_->tunnelingConfigHelper()) {
+    ASSERT(on_high_watermark_called_);
```
Could consider putting the logic of the if into the ASSERT.

e.g. `ASSERT(parent_->config-_->tunnelingConfigHelper() || (!parent_->config_->tunnelingConfigHelper() && on_high_watermark_called_))` which would ensure the entire check is compiled out in optimized mode (as currently done)

ohadvano replied:
done

KBaichoo reviewed (commented):
Nice work! Definitely on track.

ohadvano commented on envoy/network/connection.h:
```diff
@@ -84,6 +84,12 @@ class Connection : public Event::DeferredDeletable,
   enum class State { Open, Closing, Closed };
 
+  enum class ReadDisableStatus {
+    NoTransition,
```
got it, I'm not sure ``StillReadEnabled`` makes sense though - the read counter is unsigned int and always positive, so if you call ``readDisable(false)`` when the counter equals 0 that's a bug, so AFAIU you either don't call ``readDisable(false)`` if it's already enabled, or you call ``readDisable(true)`` which would be either transitioning to disable or ``StillReadDisable``. wdyt?

ohadvano replied:
I guess that your suggestion could also work, but can you please elaborate the concern about the current approach?

soulxu replied:
I don't have a strong reason for those two options. I think both are ok. Just looking at the discussion about those enum definitions, then this idea is coming up, just utilized the existing interface and return a simple bool here. So just trying to provide one more option if the discussion is going hard here. Not a comment for a required change :)

ohadvano commented:
/assign-from @envoyproxy/senior-maintainers

ohadvano commented:
/unassign @lizan 
currently out

ohadvano commented:
/assign-from @envoyproxy/senior-maintainers

ohadvano commented:
@KBaichoo I added tests, and I think it's ready for review
Could you please do another pass and also assign senior which is not Lizan, as he's currently out? Tried unassigning and re-auto assign, but it chose him again

ohadvano commented:
/retest

KBaichoo commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -293,11 +293,14 @@ void Filter::readDisableDownstream(bool disable) {
-  read_callbacks_->connection().readDisable(disable);
 
-  if (disable) {
+  Network::Connection::ReadDisableStatus read_disable_status =
```
const

KBaichoo reviewed (commented):
Can you add an integration test to reproduce the problem that motivated this work so we can have a regression test? https://github.com/envoyproxy/envoy/issues/27516

Otherwise looks great!

/assign @RyanTheOptimist 
/wait for integration test

RyanTheOptimist commented on envoy/network/connection.h:
```diff
@@ -186,7 +193,7 @@ class Connection : public Event::DeferredDeletable,
    * readDisable(false);  // Marks the connection as unblocked, so resumes reading.
    */
-  virtual void readDisable(bool disable) PURE;
+  virtual ReadDisableStatus readDisable(bool disable) PURE;
```
Nit: please document the `@return` value

ohadvano replied:
done

RyanTheOptimist reviewed (commented):
Agreed about the integration test. Otherwise, one nit.

ohadvano commented:
/retest

ohadvano commented:
@KBaichoo, @RyanTheOptimist I addressed your comments. Regarding the integration tests - I couldn't figure out how to simulate an integration test that hits those callbacks twice. It seems like ``FakeUpstream`` watermark callbacks are null-op. If you have an idea or can help with this, I'll be happy to follow your guidance, otherwise - maybe we should add a backlog to add more comprehensive watermark tests?

KBaichoo commented:
I think you could achieve this by following something similar to tests in `test/integration/buffer_accounting_integration_test.cc` by using the  socket swap.

/wait

ohadvano commented:
Thanks for the reference, however it still seems to me that even if I use this approach, the ``onAboveWriteBufferHighWatermark`` will not propagate to the ``tcp_proxy`` filter's callback, because of the null-op here: 
https://github.com/envoyproxy/envoy/blob/1ef7992acc2259b208b1c8c5ea84dcd33f2b072d/test/integration/fake_upstream.h#L238

Unless I misunderstood something, it looks like it would require some refactoring to get this done, could you please share your insight or extra details about what you had in mind for testing this?

KBaichoo commented:
See: https://github.com/envoyproxy/envoy/blob/1ef7992acc2259b208b1c8c5ea84dcd33f2b072d/test/integration/buffer_accounting_integration_test.cc#L650 as an example.

I think the original problem was TCP proxy <---> HTTP2 upstream cluster; we could size the HTTP2 config / upstream connection buffer to be < downstream connection buffer and use the socket swap to make writes to upstream fail.

We should then be able to get disable from the upstream connection and upstream codec.

ohadvano commented:
I do think it's possible to get the upstream request to disable and call ``onAboveWriteBufferHighWatermark``, but how's it going to propagate to the ``tcp_proxy::onAboveWriteBufferHighWatermark``? Because it' currently null-op, as I see it the fake upsream is unaware of the callbacks defined by the ``tcp_proxy`` side, and I'm not sure if it's currently possible to pass those callbacks

KBaichoo commented:
> as I see it the fake upstream is unaware of the callbacks defined by the tcp_proxy side, and I'm not sure if it's currently possible to pass those callbacks

Recommend running the tests I've linked above and you should see the writes don't reach the upstream (they end up being stuck in Envoy due to socket swap injecting errors) which should trigger the watermarking callbacks.

ohadvano commented:
Ok thanks I'll try this. Just to make sure I understand you, in ``tcp_proxy`` there are two callbacks sets, downstream:
https://github.com/envoyproxy/envoy/blob/1ef7992acc2259b208b1c8c5ea84dcd33f2b072d/source/common/tcp_proxy/tcp_proxy.cc#L309

And upstream:
https://github.com/envoyproxy/envoy/blob/1ef7992acc2259b208b1c8c5ea84dcd33f2b072d/source/common/tcp_proxy/tcp_proxy.cc#L335

Are you saying that this approach should trigger the ``Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark()``?

KBaichoo commented:
Yes I think we should be able to trigger the `Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark` as done in your production case by:
1) Having the upstream network connection buffer reach watermark
2) Having the HTTP2 codec watermark

Thanks for giving this a shot, it can be tricky.

ohadvano commented:
@KBaichoo, so I did try to get this done with the ``SocketInterfaceSwap``, but after a lot of variations I still couldn't get that callback to fire twice.
With the following version for example, the callback isn't called at all.

```
TEST_P(TcpTunnelingIntegrationWatermarkTest, testemp) {
  initialize();
  setUpConnection(fake_upstream_connection_);
  write_matcher_->setDestinationPort(fake_upstreams_[0]->localAddress()->ip()->port());
  write_matcher_->setWriteReturnsEgain();

  ASSERT_TRUE(tcp_client_->write(std::string(2000, 'a'), false));
  buffer_factory_->waitUntilTotalBufferedExceeds(1000);

  ASSERT_TRUE(fake_upstream_connection_->close());
  tcp_client_->close();
}
```

With the following variation, the upstream isn't even set up, because response headers are not set (need to call setUpConnection first that also encodes back the response ``CONNECT`` headers):

```
TEST_P(TcpTunnelingIntegrationWatermarkTest, testemp) {
  initialize();
  write_matcher_->setDestinationPort(fake_upstreams_[0]->localAddress()->ip()->port());
  write_matcher_->setWriteReturnsEgain();
  setUpConnection(fake_upstream_connection_);

  ASSERT_TRUE(tcp_client_->write(std::string(2000, 'a'), false));
  buffer_factory_->waitUntilTotalBufferedExceeds(1000);

  ASSERT_TRUE(fake_upstream_connection_->close());
  tcp_client_->close();
}
```

The closest I could get was with this variation, where ``setBufferLimits`` did the trick and actually didn't require the ``SocketInterfaceSwap``.
However, with this version the callback was called only once, when needed twice.

```
TEST_P(TcpTunnelingIntegrationWatermarkTest, testemp) {
  config_helper_.setBufferLimits(10, 10000);
  initialize();
  setUpConnection(fake_upstream_connection_);

  ASSERT_TRUE(tcp_client_->write(std::string(1000, 'a'), false));

  ASSERT_TRUE(fake_upstream_connection_->close());
  tcp_client_->close();
}
```

``TcpTunnelingIntegrationWatermarkTest`` is a test class that is the same as ``Http2OverloadManagerIntegrationTest`` only it also has the ``tcp_proxy`` listener with TCP Tunneling enabled

Any idea what else can be done? Right now I'm feeling a bit stuck on this.

ohadvano commented:
ping @KBaichoo

KBaichoo reviewed (commented):
Thank you for working on this. I think you're really close to cracking the reproduction!

If we go with your third attempt: I think all you might need to do is to configure [typed_extension_protocol_options](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto) to specify HTTP2 [initial stream window size](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-msg-config-core-v3-http2protocoloptions) to 65535 and bump up the downstream window to something larger like 128k. This is where the socket swap would be handy to ensure you are able to reach the watermark for the upstream protocol reliably.

ohadvano commented:
Thanks for guiding this @KBaichoo. I added an integration test to validate this case, and validated that with the previous ``ASSERT`` call the tests are failing due to that assert call hit.

ohadvano commented:
/retest

KBaichoo commented on test/integration/buffer_accounting_integration_test.cc:
```diff
@@ -606,6 +607,69 @@ TEST_P(ProtocolsBufferWatermarksTest, ResettingStreamUnregistersAccount) {
+
+  write_matcher_->setWriteReturnsEgain();
+  ASSERT_TRUE(tcp_client_->write(std::string(524288, 'a'), false));
+  tcp_client_->close();
```
I think we should wait here on the stat being incremented that there was a single read disable stat increment to the proxy.

ohadvano replied:
Thanks, makes sense. Added a check to equal 1. I'm not sure if there's a wait to check it doesn't reach 2 unless we make the test duration longer due to timeout, correct me if I'm wrong.

KBaichoo replied:
Nope there's no wait for it to not change (and that would just be effectively a sleep wasting CI time), thanks for checking that it breaks if we do.

KBaichoo commented on test/integration/buffer_accounting_integration_test.cc:
```diff
@@ -606,6 +607,69 @@ TEST_P(ProtocolsBufferWatermarksTest, ResettingStreamUnregistersAccount) {
+                     testing::Bool()),
+    protocolTestParamsAndBoolToString);
+
+TEST_P(TcpTunnelingWatermarkIntegrationTest, MultipleReadDisableCalls) {
```
nit: consider naming this `CanHandleMultipleReadDisableCalls` or `DoesNotAssertWhenMultipleReadDisableCalls` or `MultipleReadDisableCallsIncrementsStatsOnce` or similar.

ohadvano replied:
done

KBaichoo commented on test/integration/buffer_accounting_integration_test.cc:
```diff
@@ -606,6 +607,69 @@ TEST_P(ProtocolsBufferWatermarksTest, ResettingStreamUnregistersAccount) {
+        });
+  }
+
+  IntegrationTcpClientPtr tcp_client_;
```
member should be protected

ohadvano replied:
done

KBaichoo reviewed (commented):
Great work! Thanks for iterating on this. Looks good to me otherwise!

KBaichoo commented:
/wait

KBaichoo commented:
/retest

ohadvano commented:
/retest

ohadvano commented:
It seems like the CI is consecutively failing on ``example_config_test`` although it passed before and I'm not really sure it has anything to do with this PR changes. Any idea what's going on?

ohadvano commented:
/retest

ohadvano commented:
@phlax hopefully you could help here, how can the Verify/examples step retried?
